File: ipfs_accelerate_js/src/utils/webgpu_4bit_kernels.ts
// FI: any;
 * Convert: any;
 * Conversi: any;
 * Th: any;
 * Conversi: any;
 */;

import { HardwareAbstract: any;

// WebG: any;
/** WebG: any;

Th: any;

These kernels are designed to work with the WebGPU quantization system for) {
1: a: any;
2: a: any;
3: a: any;

Implementation Notes) {
- WG: any;
- Pyth: any;
- WebG: any;

impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
// Configu: any;
logging.basicConfig(level=logging.INFO, format) { any) { any) { any: any: any: any: any = '%(asctime: a: any;'
logger: any: any = logg: any;
// W: any;
MATRIX_MUL_4BIT_SHADER) { any: any: any: any = /** // Web: any;
struct Matrix4BitData ${$1};

struct InputMatrix ${$1};

struct OutputMatrix ${$1};

@group(0: a: any;
@group(0: a: any;
@group(0: a: any;

// Help: any;
fn unpack_4bit(packed: u32, index: u32) -> i32 ${$1}

@compute @workgroup_size(16: a: any;
fn main(@builtin(global_invocation_id: any) global_id: vec3<u32>) {let row: any: any: any: any: any: any = global: any;
  let col: any: any: any: any: any: any = global: any;}
  // Che: any;
  if ((((row >= outputMatrix.rows || col >= outputMatrix.cols) { ${$1}
  
  var sum) { f32) { any) { any) { any) { any) { any: any = 0: a: an: any;
  
  // Compu: any;
  for (((((((var k) { u32) { any) { any) { any) { any) { any: any: any: any: any: any: any = 0; k: a: an: any; k: any: any: any: any = k + 2) {// G: any;
    let input_value: any: any: any: any: any: any = inputMat: any;}
    // Calcula: any;
    let packed_idx: any: any: any: any: any: any = (row * weightMat: any;
    let sub_idx: any: any: any: any: any: any = (row * weightMat: any;
    
    // G: any;
    let packed_weight: any: any: any: any: any: any = weightMat: any;
    
    // Unpa: any;
    let weight1: any: any: any: any: any: any = unpack_4: any;
    
    // Dequanti: any;
    let dequantized_weight1: any: any: any: any: any: any = f: an: any;
    
    // Multip: any;
    sum: any: any: any: any: any: any = s: an: any;
    
    // I: an: any;
    if (((((((k + 1 < inputMatrix.cols) { ${$1}
  
  // Write) { an) { an: any;
  outputMatrix.data[row * outputMatrix.cols + col] = su) {any;} */;

// WGS) { an: any;
ATTENTION_4BIT_SHADER) { any) { any) { any) { any) { any: any: any: any = /** // WebG: any;
struct Matrix4BitData ${$1};

struct FloatMatrix ${$1};

struct AttentionParams ${$1};

@group(0) { any) { @binding(0: any) var<storage, read> query_weights) { Matrix4BitDat) { a: an: any;
@group(0: a: any;
@group(0: a: any;
@group(0: a: any;
@group(0: a: any;
@group(0: a: any;

// Helper functions for ((((((4-bit operations (same as matrix mul) {
fn unpack_4bit(packed) { any)) { any { u32, index) { u32) -> i32 ${$1}

fn dequantize(packed_idx) { u32, sub_idx) { u32, matrix: Matrix4BitData) -> f32 ${$1}

// Speci: any;
@compute @workgroup_size(8) { any, 8, 1: any) {
fn main(@builtin(global_invocation_id: any) global_id) { vec3<u32>) {
  let batch_idx) {any: any: any: any: any: any = global: any;
  let seq_pos: any: any: any: any: any: any = global: any;
  let head_idx: any: any: any: any: any: any = global: any;
  let head_pos: any: any: any: any: any: any = global: any;}
  // Che: any;
  if (((((((batch_idx >= params.batch_size || head_idx >= params.num_heads) { ${$1}
  
  // Calculate) { an) { an: any;
  let input_base) { any) { any) { any) { any: any: any = batch_: any;
  
  // Calcula: any;
  var q_value: f32: any: any: any: any: any: any = 0: a: an: any;
  var k_value: f32: any: any: any: any: any: any = 0: a: an: any;
  var v_value: f32: any: any: any: any: any: any = 0: a: an: any;
  
  // Proje: any;
  for (((((((var i) { u32) { any) { any) { any) { any) { any: any: any: any: any: any: any = 0; i: a: an: any; i++) ${$1}
  
  // Wri: any;
  // I: an: any;
  let output_idx: any: any: any: any: any: any = batch_: any;
          
  attention_output.data[output_idx] = q_va: any;
} */;

class $1 extends $2 {/** Implemen: any;
  i: an: any;
  
  function this( this: any:  any: any): any {  any: any): any {: any { any, 
        $1) {: any { boolean: any: any: any = tr: any;
        $1: boolean: any: any = tr: any;
    /** Initiali: any;
    
    A: any;
      use_mixed_precis: any;
      optimize_attent: any;
    this.use_mixed_precision = use_mixed_precis: any;
    this.optimize_attention = optimize_attent: any;
    
    // Performan: any;
    this.performance_stats = ${$1}
    
    logg: any;
    logg: any;
    logg: any;
  
  $1($2): $3 {/** G: any;
    return MATRIX_MUL_4BIT_SHADER}
  $1($2) {) { $3 {/** G: any;
    return ATTENTION_4BIT_SHADER}
  function this( this: any:  any: any): any {  any: any): any {: any { any, 
        $1) {: any { Reco: any;
    /** Simula: any;
    
    A: any;
      weights_4: any;
      input_activati: any;
      
    Retu: any;
      Matr: any;
    start_time: any: any: any = ti: any;
    
    // Extra: any;
    quantized_data: any: any = (weights_4bit["data"] !== undefin: any;"
    if ((((((($1) {throw new) { an) { an: any;
    weight_shape) { any) { any = (weights_4bit["shape"] !== undefine) { an: any;"
    weight_rows, weight_cols: any: any: any = weight_sh: any;
    
    // G: any;
    quant_params: any: any = (weights_4bit["params"] !== undefined ? weights_4bit["params"] : {});"
    scale: any: any = (quant_params["scale"] !== undefin: any;"
    zero_point: any: any = (quant_params["zero_point"] !== undefin: any;"
    bits: any: any = (weights_4bit["bits"] !== undefin: any;"
    
    // Che: any;
    input_shape: any: any: any = input_activatio: any;
    if (((((($1) {
      input_activations) {any = input_activations) { an) { an: any;
      input_shape) { any) { any: any = input_activatio: any;}
    input_rows, input_cols: any: any: any = input_sh: any;
    
    // Veri: any;
    if (((((($1) {throw new) { an) { an: any;
    output_shape) { any) { any = (input_rows) { a: any;
    output: any: any = np.zeros(output_shape: any, dtype: any: any: any = n: an: any;
    
    // Unpa: any;
    if (((((($1) {
      // Unpack) { an) { an: any;
      import {* a) { an: any;
      quantizer) {any = WebGPUQuantiz: any;
      unpacked_weights) { any: any = quantiz: any;}
      // Calcula: any;
      num_elements: any: any: any = weight_ro: any;
      ;
      // Resha: any;
      if (((((($1) { ${$1} else { ${$1} else {// For non-4-bit weights, fallback to standard matmul}
      dequantized_weights) { any) { any) { any = (weights_4bit["data"] !== undefined) { an) { an: any;"
      output: any: any = n: an: any;
    
    // Reco: any;
    matmul_time: any: any: any = (time.time() - start_ti: any;
    this.performance_stats["matmul_time_ms"] = matmul_t: any;"
    
    retu: any;
  
  functi: any;
          $1): any { Reco: any;
          $1: Reco: any;
          $1: Reco: any;
          input_activati: any;
          $1: numb: any;
          $1: numb: any;
    /** Simula: any;
    
    A: any;
      query_weights_4: any;
      key_weights_4: any;
      value_weights_4: any;
      input_activati: any;
      num_he: any;
      head_s: any;
      
    Retu: any;
      Attenti: any;
    start_time: any: any: any = ti: any;
    
    // Comm: any;
    batch_size, seq_length: any, hidden_size: any: any: any = input_activatio: any;
    
    // Calcula: any;
    query: any: any = th: any;
    key: any: any = th: any;
    value: any: any = th: any;
    
    // Resha: any;
    query: any: any = que: any;
    key: any: any = k: any;
    value: any: any = val: any;
    
    // Transpo: any;
    query) { any) { any = query.transpose(0: any, 2, 1: any, 3) {  // [batch, num_he: any;
    key: any: any = k: any;
    value: any: any = val: any;
    
    // Calcula: any;
    attention_scores: any: any = n: an: any;
    
    // Sca: any;
    attention_scores: any: any = attention_scor: any;
    
    // App: any;
    attention_probs: any: any = np.exp(attention_scores - np.max(attention_scores: any, axis: any: any = -1, keepdims: any: any: any = tr: any;
    attention_probs: any: any = attention_probs / np.sum(attention_probs: any, axis: any: any = -1, keepdims: any: any: any = tr: any;
    
    // Calcula: any;
    context: any: any = n: an: any;
    
    // Transpo: any;
    context: any: any = conte: any;
    
    // Resha: any;
    context: any: any = conte: any;
    
    // Reco: any;
    attention_time: any: any: any = (time.time() - start_ti: any;
    this.performance_stats["attention_time_ms"] = attention_t: any;"
    
    retu: any;
  ;
  function this(this:  any:  any: any:  any: any): any -> Dict[str, float]) {
    /** G: any;
    retu: any;


$1($2) {/** Examp: any;
  // Crea: any;
  input_size: any: any: any = 7: an: any;
  hidden_size: any: any: any = 3: any;}
  // Crea: any;
  input_activations: any: any = n: an: any;
  
  // Crea: any;
  weights: any: any = n: an: any;
  
  // Initiali: any;
  kernel: any: any: any = WebGPU4BitKerne: any;
  
  // Quanti: any;
  import {* a: an: any;
  quantizer: any: any: any: any: any: any = WebGPUQuantizer(default_bits=4);
  ;
  // Conve: any;
  weights_4bit: any: any = {
    "data": np.random.randparseInt(-8, 8: any, size: any: any: any = (hidden_size * input_si: any;"
    "shape": (hidden_size: a: any;"
    "bits": 4: a: any;"
    "params": ${$1}"
  
  // Measu: any;
  start_time: any: any: any: any: any: any: any = ti: any;
  fp32_result: any: any = n: an: any;
  fp32_time: any: any: any = (time.time() - start_ti: any;
  
  // Measu: any;
  start_time: any: any: any = ti: any;
  b4_result: any: any = kern: any;
  b4_time: any: any: any = (time.time() - start_ti: any;
  
  // Pri: any;
  conso: any;
  conso: any;
  conso: any;
  conso: any;
  
  // Pri: any;
  fp32_memory: any: any: any = input_si: any;
  int4_memory: any: any = input_size * hidden_size // Math.floor(2 / 4) bits per value: any: any: any = 1: a: any;
  
  fp32_memory_mb: any: any: any = fp32_memo: any;
  int4_memory_mb: any: any = int4_mem: any;
if (((($1) {;
  example_4bit_matmul) { an) { an) { an: any;

File: tests/test/fixed_web_platform/webgpu_4bit_kernels.py
#!/usr/bin/env python3
"""
WebGPU 4-bit Matrix Multiplication Kernels

This module provides optimized WebGPU compute shader implementations for 4-bit matrix 
operations, enabling high-performance LLM inference in browser environments with 
significantly reduced memory usage.

These kernels are designed to work with the WebGPU quantization system for:
1. Efficient matrix multiplication with 4-bit weights
2. Mixed-precision operations (4-bit weights with higher precision activations)
3. Optimized attention calculation for transformer models

Implementation Notes:
- WGSL (WebGPU Shading Language) shaders for hardware acceleration
- Python simulation for validation and testing
- WebGPU-specific kernel optimizations
"""

import os
import sys
import time
import json
import logging
import numpy as np
from pathlib import Path
from typing import Dict, List, Any, Optional, Union, Tuple, Callable

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger("webgpu_4bit_kernels")

# WGSL shader for 4-bit matrix multiplication
MATRIX_MUL_4BIT_SHADER = """
// WebGPU 4-bit matrix multiplication compute shader
struct Matrix4BitData {
    // Packed 4-bit values (each uint32 contains 8 values)
    data: array<u32>,
    // Original matrix dimensions
    rows: u32,
    cols: u32,
    // Quantization parameters
    scale: f32,
    zero_point: i32,
};

struct InputMatrix {
    data: array<f32>,
    rows: u32,
    cols: u32,
};

struct OutputMatrix {
    data: array<f32>,
    rows: u32,
    cols: u32,
};

@group(0) @binding(0) var<storage, read> weightMatrix: Matrix4BitData;
@group(0) @binding(1) var<storage, read> inputMatrix: InputMatrix;
@group(0) @binding(2) var<storage, write> outputMatrix: OutputMatrix;

// Helper function to unpack 4-bit values
fn unpack_4bit(packed: u32, index: u32) -> i32 {
    let shift = (index % 8) * 4;
    let mask = 0xF << shift;
    let value = (packed & mask) >> shift;
    
    // Convert to signed 4-bit integer (-8 to 7)
    return i32(value) - 8;
}

@compute @workgroup_size(16, 16)
fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {
    let row = global_id.x;
    let col = global_id.y;
    
    // Check if we're within bounds
    if (row >= outputMatrix.rows || col >= outputMatrix.cols) {
        return;
    }
    
    var sum: f32 = 0.0;
    
    // Compute the dot product
    for (var k: u32 = 0; k < inputMatrix.cols; k = k + 2) {
        // Get the input activation value
        let input_value = inputMatrix.data[col * inputMatrix.cols + k];
        
        // Calculate packed 4-bit weight index
        let packed_idx = (row * weightMatrix.cols + k) / 8;
        let sub_idx = (row * weightMatrix.cols + k) % 8;
        
        // Get the packed weight value
        let packed_weight = weightMatrix.data[packed_idx];
        
        // Unpack first 4-bit weight
        let weight1 = unpack_4bit(packed_weight, sub_idx);
        
        // Dequantize the weight
        let dequantized_weight1 = f32(weight1 - weightMatrix.zero_point) * weightMatrix.scale;
        
        // Multiply and accumulate
        sum = sum + dequantized_weight1 * input_value;
        
        // If we have another weight (and haven't gone out of bounds)
        if (k + 1 < inputMatrix.cols) {
            // Get the next input value
            let input_value2 = inputMatrix.data[col * inputMatrix.cols + k + 1];
            
            // Calculate next packed 4-bit weight index
            let packed_idx2 = (row * weightMatrix.cols + k + 1) / 8;
            let sub_idx2 = (row * weightMatrix.cols + k + 1) % 8;
            
            // Get the packed weight value (might be the same as the first one)
            let packed_weight2 = weightMatrix.data[packed_idx2];
            
            // Unpack second 4-bit weight
            let weight2 = unpack_4bit(packed_weight2, sub_idx2);
            
            // Dequantize the weight
            let dequantized_weight2 = f32(weight2 - weightMatrix.zero_point) * weightMatrix.scale;
            
            // Multiply and accumulate
            sum = sum + dequantized_weight2 * input_value2;
        }
    }
    
    // Write output
    outputMatrix.data[row * outputMatrix.cols + col] = sum;
}
"""

# WGSL shader for attention with 4-bit weights
ATTENTION_4BIT_SHADER = """
// WebGPU 4-bit attention compute shader optimized for transformer models
struct Matrix4BitData {
    data: array<u32>,
    rows: u32,
    cols: u32,
    scale: f32,
    zero_point: i32,
};

struct FloatMatrix {
    data: array<f32>,
    rows: u32,
    cols: u32,
};

struct AttentionParams {
    head_size: u32,
    num_heads: u32,
    seq_length: u32,
    batch_size: u32,
    scale: f32,
};

@group(0) @binding(0) var<storage, read> query_weights: Matrix4BitData;
@group(0) @binding(1) var<storage, read> key_weights: Matrix4BitData;
@group(0) @binding(2) var<storage, read> value_weights: Matrix4BitData;
@group(0) @binding(3) var<storage, read> input_data: FloatMatrix;
@group(0) @binding(4) var<storage, write> attention_output: FloatMatrix;
@group(0) @binding(5) var<uniform> params: AttentionParams;

// Helper functions for 4-bit operations (same as matrix mul)
fn unpack_4bit(packed: u32, index: u32) -> i32 {
    let shift = (index % 8) * 4;
    let mask = 0xF << shift;
    let value = (packed & mask) >> shift;
    return i32(value) - 8;
}

fn dequantize(packed_idx: u32, sub_idx: u32, matrix: Matrix4BitData) -> f32 {
    let packed = matrix.data[packed_idx];
    let quant_value = unpack_4bit(packed, sub_idx);
    return f32(quant_value - matrix.zero_point) * matrix.scale;
}

// Special compute shader for self-attention with 4-bit weights
@compute @workgroup_size(8, 8, 1)
fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {
    let batch_idx = global_id.x / params.seq_length;
    let seq_pos = global_id.x % params.seq_length;
    let head_idx = global_id.y / params.head_size;
    let head_pos = global_id.y % params.head_size;
    
    // Check bounds
    if (batch_idx >= params.batch_size || head_idx >= params.num_heads) {
        return;
    }
    
    // Calculate input index base
    let input_base = batch_idx * params.seq_length * input_data.cols + seq_pos * input_data.cols;
    
    // Calculate QKV projections with 4-bit weights
    var q_value: f32 = 0.0;
    var k_value: f32 = 0.0;
    var v_value: f32 = 0.0;
    
    // Project input to Q, K, V (dot product with 4-bit weights)
    for (var i: u32 = 0; i < input_data.cols; i++) {
        let input_val = input_data.data[input_base + i];
        
        // Query projection
        let q_packed_idx = (head_idx * params.head_size + head_pos) * query_weights.cols + i / 8;
        let q_sub_idx = i % 8;
        q_value += input_val * dequantize(q_packed_idx, q_sub_idx, query_weights);
        
        // Key projection
        let k_packed_idx = (head_idx * params.head_size + head_pos) * key_weights.cols + i / 8;
        let k_sub_idx = i % 8;
        k_value += input_val * dequantize(k_packed_idx, k_sub_idx, key_weights);
        
        // Value projection
        let v_packed_idx = (head_idx * params.head_size + head_pos) * value_weights.cols + i / 8;
        let v_sub_idx = i % 8;
        v_value += input_val * dequantize(v_packed_idx, v_sub_idx, value_weights);
    }
    
    // Write projected values to output (simplified attention)
    // In a full implementation, we'd compute attention scores, softmax, etc.
    let output_idx = batch_idx * params.num_heads * params.seq_length * params.head_size +
                     head_idx * params.seq_length * params.head_size +
                     seq_pos * params.head_size +
                     head_pos;
                     
    attention_output.data[output_idx] = q_value * k_value * v_value * params.scale;
}
"""

class WebGPU4BitKernels:
    """
    Implements optimized WebGPU compute shader kernels for 4-bit operations.
    
    This class provides a Python simulation of how 4-bit operations would be 
    implemented in WebGPU, as well as the actual WGSL shader code that would run
    in a browser environment.
    """
    
    def __init__(self, 
                 use_mixed_precision: bool = True,
                 optimize_attention: bool = True):
        """
        Initialize WebGPU 4-bit kernels.
        
        Args:
            use_mixed_precision: Whether to use mixed precision (16-bit activations)
            optimize_attention: Whether to use attention-specific optimizations
        """
        self.use_mixed_precision = use_mixed_precision
        self.optimize_attention = optimize_attention
        
        # Performance tracking
        self.performance_stats = {
            "matmul_time_ms": 0.0,
            "attention_time_ms": 0.0,
            "inference_time_ms": 0.0,
            "memory_usage_mb": 0.0
        }
        
        logger.info(f"Initialized WebGPU 4-bit kernels")
        logger.info(f"Mixed precision: {use_mixed_precision}")
        logger.info(f"Optimized attention: {optimize_attention}")
    
    def get_matmul_shader_code(self) -> str:
        """Get the WGSL shader code for 4-bit matrix multiplication."""
        return MATRIX_MUL_4BIT_SHADER
    
    def get_attention_shader_code(self) -> str:
        """Get the WGSL shader code for 4-bit attention."""
        return ATTENTION_4BIT_SHADER
    
    def matmul_4bit(self, 
                   weights_4bit: Dict[str, Any], 
                   input_activations: np.ndarray) -> np.ndarray:
        """
        Simulate 4-bit WebGPU matrix multiplication.
        
        Args:
            weights_4bit: 4-bit quantized weights with quantization parameters
            input_activations: Input activations in fp32 or fp16
            
        Returns:
            Matrix multiplication result
        """
        start_time = time.time()
        
        # Extract quantized weights and parameters
        quantized_data = weights_4bit.get("data")
        if quantized_data is None:
            raise ValueError("Weights must be quantized with quantize_model")
        
        # Get shape information
        weight_shape = weights_4bit.get("shape", (0, 0))
        weight_rows, weight_cols = weight_shape
        
        # Get quantization parameters
        quant_params = weights_4bit.get("params", {})
        scale = quant_params.get("scale", 1.0)
        zero_point = quant_params.get("zero_point", 0)
        bits = weights_4bit.get("bits", 4)
        
        # Check input dimensions
        input_shape = input_activations.shape
        if len(input_shape) != 2:
            input_activations = input_activations.reshape(-1, input_shape[-1])
            input_shape = input_activations.shape
        
        input_rows, input_cols = input_shape
        
        # Verify dimensions
        if weight_cols != input_cols:
            raise ValueError(f"Incompatible dimensions: weight_cols={weight_cols}, input_cols={input_cols}")
        
        # Allocate output tensor
        output_shape = (input_rows, weight_rows)
        output = np.zeros(output_shape, dtype=np.float32)
        
        # Unpack 4-bit weights
        if bits == 4:
            # Unpack 4-bit weights if needed
            from ..webgpu_quantization import WebGPUQuantizer
            quantizer = WebGPUQuantizer()
            unpacked_weights = quantizer._unpack_4bit_values(quantized_data)
            
            # Calculate number of elements
            num_elements = weight_rows * weight_cols
            
            # Reshape to original shape, handling potential trimming
            if len(unpacked_weights) >= num_elements:
                unpacked_weights = unpacked_weights[:num_elements].reshape(weight_shape)
            else:
                # Pad if we don't have enough elements
                padding = np.zeros(num_elements - len(unpacked_weights), dtype=unpacked_weights.dtype)
                unpacked_weights = np.concatenate([unpacked_weights, padding]).reshape(weight_shape)
            
            # Dequantize weights
            dequantized_weights = (unpacked_weights.astype(np.float32) - zero_point) * scale
            
            # Optimized matrix multiplication
            output = np.matmul(input_activations, dequantized_weights.T)
        else:
            # For non-4-bit weights, fallback to standard matmul
            dequantized_weights = weights_4bit.get("data")
            output = np.matmul(input_activations, dequantized_weights.T)
        
        # Record matmul time
        matmul_time = (time.time() - start_time) * 1000
        self.performance_stats["matmul_time_ms"] = matmul_time
        
        return output
    
    def attention_4bit(self,
                      query_weights_4bit: Dict[str, Any],
                      key_weights_4bit: Dict[str, Any],
                      value_weights_4bit: Dict[str, Any],
                      input_activations: np.ndarray,
                      num_heads: int,
                      head_size: int) -> np.ndarray:
        """
        Simulate 4-bit WebGPU attention operation.
        
        Args:
            query_weights_4bit: 4-bit quantized query weights with parameters
            key_weights_4bit: 4-bit quantized key weights with parameters
            value_weights_4bit: 4-bit quantized value weights with parameters
            input_activations: Input activations in fp32 or fp16
            num_heads: Number of attention heads
            head_size: Size of each attention head
            
        Returns:
            Attention output
        """
        start_time = time.time()
        
        # Common parameters
        batch_size, seq_length, hidden_size = input_activations.shape
        
        # Calculate Q, K, V projections using 4-bit matmul
        query = self.matmul_4bit(query_weights_4bit, input_activations.reshape(-1, hidden_size))
        key = self.matmul_4bit(key_weights_4bit, input_activations.reshape(-1, hidden_size))
        value = self.matmul_4bit(value_weights_4bit, input_activations.reshape(-1, hidden_size))
        
        # Reshape projections
        query = query.reshape(batch_size, seq_length, num_heads, head_size)
        key = key.reshape(batch_size, seq_length, num_heads, head_size)
        value = value.reshape(batch_size, seq_length, num_heads, head_size)
        
        # Transpose for attention
        query = query.transpose(0, 2, 1, 3)  # [batch, num_heads, seq_len, head_size]
        key = key.transpose(0, 2, 3, 1)      # [batch, num_heads, head_size, seq_len]
        value = value.transpose(0, 2, 1, 3)  # [batch, num_heads, seq_len, head_size]
        
        # Calculate attention scores
        attention_scores = np.matmul(query, key)
        
        # Scale attention scores
        attention_scores = attention_scores / np.sqrt(head_size)
        
        # Apply softmax
        attention_probs = np.exp(attention_scores - np.max(attention_scores, axis=-1, keepdims=True))
        attention_probs = attention_probs / np.sum(attention_probs, axis=-1, keepdims=True)
        
        # Calculate context
        context = np.matmul(attention_probs, value)
        
        # Transpose back
        context = context.transpose(0, 2, 1, 3)
        
        # Reshape to original dimensions
        context = context.reshape(batch_size, seq_length, -1)
        
        # Record attention time
        attention_time = (time.time() - start_time) * 1000
        self.performance_stats["attention_time_ms"] = attention_time
        
        return context
    
    def get_performance_stats(self) -> Dict[str, float]:
        """Get performance statistics."""
        return self.performance_stats.copy()


def example_4bit_matmul():
    """Example demonstrating 4-bit matrix multiplication performance."""
    # Create random matrices
    input_size = 768
    hidden_size = 3072
    
    # Create random input activations
    input_activations = np.random.randn(1, 128, input_size).astype(np.float32)
    
    # Create random weights
    weights = np.random.randn(hidden_size, input_size).astype(np.float32)
    
    # Initialize 4-bit kernel
    kernel = WebGPU4BitKernels()
    
    # Quantize weights (simulate)
    from ..webgpu_quantization import WebGPUQuantizer
    quantizer = WebGPUQuantizer(default_bits=4)
    
    # Convert to 4-bit (simulate)
    weights_4bit = {
        "data": np.random.randint(-8, 8, size=(hidden_size * input_size // 2)).astype(np.int8),
        "shape": (hidden_size, input_size),
        "bits": 4,
        "params": {
            "scale": 0.01,
            "zero_point": 0
        }
    }
    
    # Measure FP32 matmul time
    start_time = time.time()
    fp32_result = np.matmul(input_activations.reshape(-1, input_size), weights.T)
    fp32_time = (time.time() - start_time) * 1000
    
    # Measure 4-bit matmul time
    start_time = time.time()
    b4_result = kernel.matmul_4bit(weights_4bit, input_activations.reshape(-1, input_size))
    b4_time = (time.time() - start_time) * 1000
    
    # Print results
    print(f"Matrix shape: {input_activations.shape} x {weights.shape}")
    print(f"FP32 matmul time: {fp32_time:.2f} ms")
    print(f"4-bit matmul time: {b4_time:.2f} ms")
    print(f"Speedup: {fp32_time / b4_time:.2f}x")
    
    # Print memory usage comparison
    fp32_memory = input_size * hidden_size * 4  # 4 bytes per float32
    int4_memory = input_size * hidden_size // 2  # 4 bits per value = 1/2 byte
    
    fp32_memory_mb = fp32_memory / (1024 * 1024)
    int4_memory_mb = int4_memory / (1024 * 1024)
    
    print(f"FP32 memory: {fp32_memory_mb:.2f} MB")
    print(f"4-bit memory: {int4_memory_mb:.2f} MB")
    print(f"Memory reduction: {(fp32_memory_mb - int4_memory_mb) / fp32_memory_mb * 100:.2f}%")


if __name__ == "__main__":
    example_4bit_matmul()